### DISCUSSION TITLE
Should we handle `{ x-frame-options: SAMEORIGIN }` header ?

### DISCUSSION THREAD

**Discussion by jjavierdguezas** [responsively-org/responsively-app#606] [General]
I found and interesting scenario related to `<iframe/>` tag and the `{'x-frame-options', 'SAMEORIGIN'}` header. 

First, when that header is set, the resource can't be loaded in an `iframe` outside of its domain. So this `iframe` is not able to display cross domain.

The easiest way I tested this was using [vscode live server](https://marketplace.visualstudio.com/items?itemName=ritwickdey.LiveServer) and [express.js](https://expressjs.com/):

`index.js` file with express server:

```js
const express = require('express')
const app = express()
const port = 2000

app.get('/', (req, res) => {
  res.setHeader('x-frame-options', 'SAMEORIGIN'); // <-- We set the header here
  res.send('<div>Hello World!</div>')
})

app.listen(port, () => {
  console.log(`Example app listening at http://localhost:${port}`)
})
```

and the `index.html` file for Live Server

```html
<!DOCTYPE html>
<html lang="en">

<head>
    <meta charset="UTF-8">
    <meta http-equiv="X-UA-Compatible" content="IE=edge">
    <meta name="viewport" content="width=device-width, initial-scale=1.0">
    <title>Document</title>
</head>

<body>
    <iframe src="http://localhost:2000" title="description"></iframe>
</body>

</html>
```

if we run `node index.js` and open `index.html` file with live server, on chrome we see this: 

![image](https://user-images.githubusercontent.com/13673443/122258174-0a3a1f00-ced1-11eb-8c14-47ff566873a0.png)

but if we open `index.html` served by Live Server in *Responsively* we get an error

![image](https://user-images.githubusercontent.com/13673443/122258322-35247300-ced1-11eb-9313-5a50350553b3.png)

Should we take this in consideration for *responsively*?

If so, an easy solution found in https://github.com/electron/electron/issues/426#issuecomment-658901422 that works well is to put the following in [`main.dev`](https://github.com/responsively-org/responsively-app/blob/dcd77826f5722bc85c113b31c3dde593d85dbb5f/desktop-app/app/main.dev.js#L416) :

```js
session.defaultSession.webRequest.onHeadersReceived(
    {urls: ['*://*/*']},
    (d, c) => {
      if (d.responseHeaders['X-Frame-Options']) {
        delete d.responseHeaders['X-Frame-Options'];
      } else if (d.responseHeaders['x-frame-options']) {
        delete d.responseHeaders['x-frame-options'];
      }

      c({cancel: false, responseHeaders: d.responseHeaders});
    }
  );
```

that intercepts requests and remove the `x-frame-options` header. That way the `<iframe>` loads just fine.

![image](https://user-images.githubusercontent.com/13673443/122259861-e677d880-ced2-11eb-9ec0-03a10039b632.png)

We could also make this a configuration like the `Disable SSL Validation` one and let the user decide this behavior.
It would be nice also, if we can display an image like chrome if the `iframe` can't be loaded instead of displaying a top-level error. That way we render the rest of the page and honor the website "wishes", but maybe that is more difficult and unnecessary

What do you think @manojVivek

**Comment by jjavierdguezas** [comment]
Also we can just change in [`Webview/index.js:244`](https://github.com/responsively-org/responsively-app/blob/dcd77826f5722bc85c113b31c3dde593d85dbb5f/desktop-app/app/components/WebView/index.js#L244) when the error is not from the main frame like this:
```diff
this.webviewRef.current.addEventListener(
      'did-fail-load',
-     ({errorCode, errorDescription}) => {
+     ({errorCode, errorDescription, isMainFrame}) => {
-        if (errorCode === -3) {
+        if (!isMainFrame || errorCode === -3) {
          // Aborted error, can be ignored
          return;
        }
        this.setState({
          errorCode,
          errorDesc: errorDescription,
        });
      }
    );
```

or even
```diff
this.webviewRef.current.addEventListener(
      'did-fail-load',
-     ({errorCode, errorDescription}) => {
+     ({errorCode, errorDescription, isMainFrame}) => {
-        if (errorCode === -3) {
+        if (errorCode === -3 || (!isMainFrame && errorCode === -27)) {
          // Aborted error, can be ignored
          return;
        }
        this.setState({
          errorCode,
          errorDesc: errorDescription,
        });
      }
    );
```

and then it will be rendere like this:
![image](https://user-images.githubusercontent.com/13673443/122263456-bf230a80-ced6-11eb-858a-959ada3a632c.png)

**Comment by manojVivek** [comment]
@jjavierdguezas  I think we shouldn't allow this, that will be a real security hole. I think this cannot be compared to `Disable SSL Validation` option. 

Disabling SSL validation puts the accessing user at vulnerability and so if the user voluntarily accepts it then it is fine. 
But in the case of `X-Frame-Options`, if we disable it that makes the accessed page vulnerable. The parent frame will then be allowed access to do stuff like injecting script into the iframe, reading the iframe content, etc, which the accessed page didn't agree to by setting that header.

I think this is more against the web standards and will open possibilities for phishing-type attacks using Responsively.

**Reply by jjavierdguezas** [reply]
I understand and agree with you @manojVivek . But should we do something about the error? If a page has an invalid iframe but the rest of the content is correct, should we show an error on the whole page (like now) or just not show the content of the iframe but render the rest of the page (like chrome does)?

![image](https://user-images.githubusercontent.com/13673443/123231969-feee8100-d4d8-11eb-90fe-174046d4be38.png)

vs

![image](https://user-images.githubusercontent.com/13673443/123232292-4d038480-d4d9-11eb-9a7e-4893a2deab5d.png)

**Comment by manojVivek** [comment]
Yeah @jjavierdguezas, your solution in https://github.com/responsively-org/responsively-app/discussions/606#discussioncomment-879581 sounds good. But will be good if it is possible to show some message inside the iframe box. What do you think?

**Reply by jjavierdguezas** [reply]
yeah, a message would be good. I'll find a way 👍🏻
